Cassandra 8928 -- added sstable downgrader

Claudenw opened the pull request:
Fix for CASSANDRA-8928

This is a work in progress and I am looking for some feedback.

This fix appears to work correctly.  But I think the placement of the v3 directory is probably not the best and perhaps should be moved under the db/compaction directory where the Downgrader code is.

Suggestions appreciated.

## The changes are:

**added** 

- db/compaction/Downgrader
- tools/StandAloneDowngrader
- getNextsstableID to ColumnFamilyStore
- DOWNGRADE_SSTABLES to OperationType
- hasMaxCompressedLength to CompressionMetadata constructor and associated calls.
- V3 to SSTableFormat.Type
- added io/sstable/format/big/v3 directory containing BigFormatV3, BigTableReaderV3, BitTableScannerV3, BitTableWriterV3

**modified**

- CompressionMetadata to skip output of maxCompressedLength if not supported

**notes**

- io/sstable/format/big/v3 classes are the V3 classes modified as necessary to run within the V4 environment.

jacek-lewandowski commented on src/java/org/apache/cassandra/db/compaction/Downgrader.java:
```diff
@@ -0,0 +1,182 @@
+
+    public static SSTableFormat.Type forWriting(String version)
+    {
+        String[] v3 =
```
v3 starts with MA

Claudenw replied:
fixed

jacek-lewandowski commented on src/java/org/apache/cassandra/db/compaction/OperationType.java:
```diff
@@ -26,6 +26,7 @@
     CLEANUP("Cleanup", true, 1),
     SCRUB("Scrub", true, 1),
     UPGRADE_SSTABLES("Upgrade sstables", true, 1),
+    DOWNGRADE_SSTABLES("Downgrade sstables", true, 1),
```
perhaps not important in this case, but we should always add enums at the end because otherwise, if they are serialized as numbers somewhere, that can break the serialization

Claudenw replied:
fixed

jacek-lewandowski commented on src/java/org/apache/cassandra/io/sstable/format/SSTableFormat.java:
```diff
@@ -43,7 +43,8 @@
         //The original sstable format
-        BIG("big", BigFormat.instance);
+        BIG("big", BigFormat.instance),
+        V3("bigv3", BigFormatV3.instance);
```
Well, this is not a different format, it is a different version

Claudenw replied:
That is true, but version and format seem to be tightly coupled, unless I am missing something.  The format of some of the files has changed.  And the writer can no longer write the earlier versions.

tjake replied:
Correct.  I think you need to better document the concept which is to "Freeze" the format from the last major release.  Similar to how C* supports the last major Messaging Version to support live upgrade and downgrade.

Claudenw committed:
Added StandaloneDowngrader and associated support files.

Added io/sstable/format/big/BigFormatV3 and associated files to write the older V3 format.  Modified /io/compress/CompressionMetadata.java to adjust writing to support older format.

patch by Claude Warren; reviewed by Jake Luciani and  Jacek Lewandowski for CASSANDRA-8928

tjake commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableScannerV3.java:
```diff
@@ -0,0 +1,497 @@
+import static org.apache.cassandra.dht.AbstractBounds.maxLeft;
+import static org.apache.cassandra.dht.AbstractBounds.minRight;
+
+public class BigTableScannerV3 implements ISSTableScanner
```
Since our contract is to always support upgrading from prior versions we should be able to re-use the `BigTableScanner` meaning no need to copy this one

tjake commented on src/java/org/apache/cassandra/io/sstable/format/big/BigFormatV3.java:
```diff
@@ -0,0 +1,309 @@
+         * have no 'static' bits caused by using the same upper bits for both bloom filter and token distribution.
+         */
+        private final boolean hasOldBfFormat;
+        BigVersion(String version)
```
The logic here can be simplified to be just the flags from the `current_version`.  Otherwise this gives the impression it can write all these versions.  I would change the comments to make this clear

Claudenw committed:
fixed issue with naming

Claudenw committed:
attempt to remove / edit columns during downgrade

Claudenw committed:
fixed checkstyle issues

Claudenw committed:
working version?

Claudenw committed:
removed misleading var settings

Claudenw commented:
@jacek-lewandowski @tjake This code is to downgrade from 4.x to 3.x  So I think it needs to be rebased to a 4.x branch.  However, I think that because it is added functionality it should be a new 4.2 not a 4.1 branch.  In any case I think I need to rebase this to 4.1 (rather than trunk).  Do you agree or should I do this differently?
